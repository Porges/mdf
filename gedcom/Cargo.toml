[package]
name = "gedcom"
version = "0.1.0"
edition = "2021"
authors = ["George Pollard <porges@porg.es>"]
license = "GPL-3.0"

[features]
default = ["miette-highlighting"]
miette-highlighting = ["miette/fancy-no-backtrace"]
legacy-encodings = ["dep:oem_cp"]

[dependencies]
ascii = "1.1.0"
encoding_rs = "0.8.34"
# TODO: update once this is merged https://github.com/zkat/miette/pull/393
miette = { git = "https://github.com/Porges/miette", branch = "invalid_unicode" }
oem_cp = { version = "2.0.0", optional = true }
owo-colors = "4.0.0"
thiserror = "1.0.63"
tracing = { version = "0.1", features = ["attributes"] }
vec1 = "1.12.1"

[dev-dependencies]
insta = { version = "1.39.0", features = ["glob"] }
kdl = "4.6.0"
