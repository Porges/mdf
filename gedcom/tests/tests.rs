use std::{convert::Infallible, path::PathBuf};

use gedcom::{Collector, GedcomError, RawLine, Sink, Sourced};
use miette::{Context, NamedSource};

#[test]
fn can_parse_allged_lines() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/gpdf/allged.ged");

    let data = std::fs::read(path).unwrap();

    let line_counter = LineCounter { count: 0 };

    let line_count = gedcom::parse_lines::<_, GedcomError>(&data, line_counter).unwrap();

    assert_eq!(line_count, 1157);
}

struct LineCounter {
    count: usize,
}

impl Sink<(Sourced<usize>, Sourced<RawLine<'_>>)> for LineCounter {
    type Output = usize;
    type Err = Infallible;

    fn consume(&mut self, _value: (Sourced<usize>, Sourced<RawLine>)) -> Result<(), Self::Err> {
        self.count += 1;
        Ok(())
    }

    fn complete(self) -> Result<Self::Output, Self::Err> {
        Ok(self.count)
    }
}

#[test]
fn produces_expected_tree() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/gpdf/allged.ged");

    let data = std::fs::read(path).unwrap();

    let tree_builder = gedcom::RecordTreeBuilder::<_, GedcomError>::new(gedcom::Collector::new());
    let result = gedcom::parse_lines::<_, GedcomError>(&data, tree_builder).unwrap();

    insta::assert_debug_snapshot!(result);
}

#[test]
fn torture_test_valid() {
    insta::glob!("tests/external/TestGED/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        let result = gedcom::validate(&data).with_context(|| format!("validating {}", filename));
        match result {
            Ok(tree) => insta::assert_debug_snapshot!(tree),
            Err(err) => insta::assert_snapshot!(format!(
                "{:?}",
                err.with_source_code(NamedSource::new(filename, data))
            )),
        };
    });
}

#[test]
fn golden_files() -> miette::Result<()> {
    miette::set_hook(Box::new(|_diag| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .terminal_links(false)
                .unicode(true)
                .color(false)
                .width(132)
                .build(),
        )
    }))?;

    insta::glob!("syntax_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap();
        let tree_builder = gedcom::RecordTreeBuilder::<_, GedcomError>::new(Collector::new());
        let result = gedcom::parse_lines::<_, GedcomError>(&data, tree_builder);
        match result {
            Ok(tree) => insta::assert_debug_snapshot!(tree),
            Err(err) => insta::assert_snapshot!(format!(
                "{:?}",
                miette::Report::new(err)
                    .with_source_code(NamedSource::new(filename.to_string_lossy(), data))
            )),
        };
    });

    insta::glob!("format_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        let result = gedcom::validate(&data).with_context(|| format!("validating {}", filename));
        match result {
            Ok(tree) => insta::assert_debug_snapshot!(tree),
            Err(err) => insta::assert_snapshot!(format!(
                "{:?}",
                err.with_source_code(NamedSource::new(filename, data))
            )),
        };
    });

    Ok(())
}
