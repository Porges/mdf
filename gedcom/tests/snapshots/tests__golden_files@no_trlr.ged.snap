---
source: gedcom/tests/tests.rs
expression: "format!(\"{:?}\", err.with_source_code(NamedSource::new(filename, data)))"
input_file: gedcom/tests/format_inputs/no_trlr.ged
---
gedcom::file_format_error::missing_trailer_record

  × validating no_trlr.ged
  ├─▶ File structure error
  ╰─▶ Missing trailer (TRLR) record
  help: this record is always required at the end of the file – GEDCOM file might be truncated?
